### PR TITLE
gh-80192: Use windows api if ssl cert verification fails

### PR DESCRIPTION
**Pull request by tobil4sk**
This avoids rejecting certificates that should be accepted, see #80192.

Here is the relevant openssl documentation: https://docs.openssl.org/master/man3/SSL_CTX_set_verify/.

Also, here are the windows api docs:
- https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-certverifycertificatechainpolicy
- https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/ns-wincrypt-cert_chain_policy_status for error status.

I've been testing manually by deleting `DigiCert Global Root G2` from `Third Party Root Certificates` in the user certificate store (this is safe, the certificate will be added back automatically), and running the following code:

```python
import http.client
conn = http.client.HTTPSConnection("www.verisign.com")
conn.request("GET", "/")
conn.close()
```

The main branch errors with:
```
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1025)
```
Whereas with this patch, the certificate is verified and the connection is established.

* Issue: gh-80192

### CODE REVIEW

**Comment by asvetlov** [comment]
Such change requires some tests at least I guess.

File: Misc/NEWS.d/next/Windows/2024-12-05-00-21-11.gh-issue-80192.28PTj4.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Fixed valid ssl certificates being rejected.
```
Comment (ZeroIntensity) [inline Misc/NEWS.d/next/Windows/2024-12-05-00-21-11.gh-issue-80192.28PTj4.rst]: Use a reference:
```suggestion
Fixed valid :mod:`ssl` certificates being rejected.
```

File: Modules/_ssl.c [hunk 2]
```diff
@@ -2996,6 +2996,119 @@ static PyType_Spec PySSLSocket_spec = {
...
+_get_cert_bytes(const X509 *cert, int *length)
+{
+    unsigned char *cert_bytes;
+    int cert_bytes_length = i2d_X509(cert, NULL);
+    if (cert_bytes_length <= 0)
+    {
+        return NULL;
+    }
+
+    cert_bytes = PyMem_RawMalloc(cert_bytes_length);
```
Comment (ZeroIntensity) [inline Modules/_ssl.c:3009]: Is it possible to use pymalloc (`PyMem_Malloc`) here? Typically, we only need `RawMalloc` if we plan on using it inside `Py_BEGIN_ALLOW_THREADS` areas.

File: Modules/_ssl.c [hunk 3]
```diff
@@ -2996,6 +2996,119 @@ static PyType_Spec PySSLSocket_spec = {
...
+_verify_callback(int preverify_ok, X509_STORE_CTX *ctx)
+{
+    int ret = 0;
+    /* windows api calls below do not check for hostname mismatch */
+    if (preverify_ok || X509_STORE_CTX_get_error(ctx) == X509_V_ERR_HOSTNAME_MISMATCH)
+    {
+        return preverify_ok;
+    }
+
+    HCERTSTORE store = CertOpenStore(CERT_STORE_PROV_MEMORY, 0, 0, CERT_STORE_DEFER_CLOSE_UNTIL_LAST_FREE_FLAG, NULL);
```
Comment (ZeroIntensity) [inline Modules/_ssl.c]: There's a lot of lines in this function that are too long (the guideline is 79 characters, per PEP-7). Could you break them up across multiple lines?

File: Modules/_ssl.c [hunk 4]
```diff
@@ -2996,6 +2996,119 @@ static PyType_Spec PySSLSocket_spec = {
...
+        break;
+    case CERT_E_EXPIRED:
+        X509_STORE_CTX_set_error(ctx, X509_V_ERR_CERT_HAS_EXPIRED);
+        break;
+    case CERT_E_INVALID_POLICY:
+        X509_STORE_CTX_set_error(ctx, X509_V_ERR_INVALID_POLICY_EXTENSION);
+        break;
+    case CERT_E_WRONG_USAGE:
+    case CERT_E_CRITICAL:
+    case CERT_E_PURPOSE:
```
Comment (ZeroIntensity) [inline Modules/_ssl.c]: There's a chance we need to use `_Py_FALLTHROUGH` here, but it looks like build isn't complaining. If it does start failing feel free to use that.

**Review by ZeroIntensity — COMMENTED** [review]
Yeah, this needs a test. FYI, `test_ssl` is failing.

File: Modules/_ssl.c (hunk 2 above)
Comment (tobil4sk) [reply Modules/_ssl.c:3009]: Using PyMem_Malloc I got the following error in my sample program:
`Fatal Python error: _PyMem_DebugMalloc: Python memory allocator called without holding the GIL`

File: Modules/_ssl.c (hunk 3 above)
Comment (tobil4sk) [reply Modules/_ssl.c]: I've fixed most violations except line 3063 which has a long constant name and is a couple characters too long, does it make more sense there to enforce the rule or to leave it?

File: Misc/NEWS.d/next/Windows/2024-12-05-00-21-11.gh-issue-80192.28PTj4.rst (hunk 1 above)
Comment (tobil4sk) [reply Misc/NEWS.d/next/Windows/2024-12-05-00-21-11.gh-issue-80192.28PTj4.rst]: Reference added

File: Modules/_ssl.c [hunk 5]
```diff
@@ -2996,6 +2996,125 @@ static PyType_Spec PySSLSocket_spec = {
...
+    case TRUST_E_CERT_SIGNATURE:
+        return X509_V_ERR_CERT_SIGNATURE_FAILURE;
+    case CERT_E_UNTRUSTEDROOT:
+        return X509_V_ERR_CERT_UNTRUSTED;
+    case CERT_E_CHAINING:
+        return X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT;
+    case CERT_E_EXPIRED:
+        return X509_V_ERR_CERT_HAS_EXPIRED;
+    case CERT_E_INVALID_POLICY:
+        return X509_V_ERR_INVALID_POLICY_EXTENSION;
```
Comment (tobil4sk) [inline Modules/_ssl.c:3038]: This patch seems to satisfy the other failing check in `test_ssl`, however, I'm not sure whether this is correct:

```suggestion
    case CERT_E_INVALID_POLICY:
        return X509_V_ERR_INVALID_CA;
```

Here are the relevant docs:

CERT_E_INVALID_POLICY:
The certificate has an invalid policy.

X509_V_ERR_INVALID_CA:
A CA certificate is invalid. Either it is not a CA or its extensions are
not consistent with the supplied purpose.


- https://docs.openssl.org/master/man3/X509_STORE_CTX_get_error/#error-codes
- https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/ns-wincrypt-cert_chain_policy_status#members

**Comment by tobil4sk** [comment]
> Such change requires some tests at least I guess.

Ok, I'll have to look into how to create a sample certificate chain that reproduces the problem.

File: Modules/_ssl.c (hunk 2 above)
Comment (ZeroIntensity) [reply Modules/_ssl.c:3009]: I figured. It might be worth adding a comment mentioning that the GIL isn't held (so no Python APIs can get called).

File: Misc/NEWS.d/next/Windows/2024-12-05-00-21-11.gh-issue-80192.28PTj4.rst (hunk 1 above)
Comment (asvetlov) [reply Misc/NEWS.d/next/Windows/2024-12-05-00-21-11.gh-issue-80192.28PTj4.rst]: I would see a broader text, e.g. the PR verifies certifications rejected by OpenSSL by additional calls to Windows specific API functions.
The exact proposed text is definitely not fine, but you have got my idea.

File: Modules/_ssl.c (hunk 3 above)
Comment (ZeroIntensity) [reply Modules/_ssl.c]: I think it's fine.

File: Modules/_ssl.c (hunk 5 above)
Comment (tobil4sk) [reply Modules/_ssl.c:3038]: It looks like test_ssl isn't failing anymore actually, however, test_poplib is failing due to a different error

File: Misc/NEWS.d/next/Windows/2024-12-05-00-21-11.gh-issue-80192.28PTj4.rst [hunk 6]
```diff
@@ -0,0 +1 @@
+Fixed valid :mod:`ssl` certificates being rejected.
```
Comment (zooba) [inline Misc/NEWS.d/next/Windows/2024-12-05-00-21-11.gh-issue-80192.28PTj4.rst:1]: Agree with Andrew's comment - this needs to be significantly expanded, because this is a very impactful change.

We also would need a section in `Doc/whatsnew/3.14.rst` explaining the impact and options for workarounds.

We _might_ also need to use a deprecation period to turn this on by default, though part of me would prefer not to.

**Comment by zooba** [comment]
I wonder if @sethmlarson has thoughts?

**Comment by woodruffw** [comment]
> I'm in favor of moving to using system certificate stores by default, but I think it should happen through the PEP process to get the API right.

Agreed with this. 

(Tangentially, I think the approach in this path is unfortunately hard to follow -- if the goal is to inject the Windows cert store into the path validation flow, then IMO that should be done by loading the store and injecting it into the OpenSSL equivalent, rather than adding a verification callback and calling the equivalent WinCrypt APIs *in situ*. The latter effectively means that two path validation flows are running at the same time!)

**Comment by sethmlarson** [comment]
> > because Windows will update the store dynamically based on validation requests. So you can't just read it into OpenSSL and be done (or this wouldn't continue being raised as a problem for the last 15+ years).
> 
> Oh, huh. Is this because they're doing a variant of the Chrome intermediate cache? I wasn't aware any OS-native TLS stacks were doing that!

This is the reason why Truststore doesn't simply load certificates into an OpenSSL store, we use Windows' validation. Doesn't macOS do the same?

**Comment by woodruffw** [comment]
> This is the reason why Truststore doesn't simply load certificates into an OpenSSL store, we use Windows' validation. Doesn't macOS do the same?

I'm not sure -- I'm aware of this technique in the browser world for handling broken/misconfigured HTTPS servers, but I wasn't aware of it being employed within the OS validator itself (I was under the impression that Chrome et al. did it by maintaining a browser-side intermediate cache, which they then fed into the system APIs, with the latter remaining stateless).

Looking at this some more, I'm not actually sure that's what's happening in the Windows case: I think the Windows system validator might be doing AIA chasing instead, i.e. fetching the intermediate from the `AuthorityInformationAccess` endpoint instead of from a local intermediate cache. That would make a lot more sense than a global browser-style intermediate cache at the OS level, but someone who knows these platform verifiers better than I do could answer definitively 😅

**Comment by zooba** [comment]
> Looking at this some more, I'm not actually sure that's what's happening in the Windows case: I think the Windows system validator might be doing AIA chasing instead, i.e. fetching the intermediate from the `AuthorityInformationAccess` endpoint instead of from a local intermediate cache.

Windows used to distribute root CA updates through Windows Update, so my assumption is that it's a replica of that process, just more finely grained. Probably also works as an alternative to CRLs.

The Windows cert store isn't just used for TLS, so the design is going to be less biased towards online scenarios, as it has to include software and driver certifications as well. Though it wouldn't surprise me if most code signing CAs are included out of the box, or else OEMs/users would have a terrible time adding drivers.

**Comment by tobil4sk** [comment]
From the other comments, it sounds like this will require a more developed approach with a proper proposal.
